chainer/functions/math/matmul.py

```diff
@@ -81,84 +82,54 @@ def check_type_forward(self, in_types):
         type_check.expect(
             a_type.dtype.kind == 'f',
             b_type.dtype.kind == 'f',
-            a_type.ndim >= 1,
-            a_type.ndim == b_type.ndim,
         )
 
-        ndim = type_check.eval(a_type.ndim)
-        if ndim == 1:
-            type_check.expect(a_type.shape == b_type.shape)
+        a_ndim = type_check.eval(a_type.ndim)
+        b_ndim = type_check.eval(b_type.ndim)
+        if a_ndim == 0 or b_ndim == 0:
+            pass
+        elif a_ndim == 1 or b_ndim == 1:
+            type_check.expect(
+                a_type.ndim == b_type.ndim,
+                a_type.shape == b_type.shape,
+            )
         else:
             a_idx = _get_check_index(self.transa, False,
                                      row_idx=-2, col_idx=-1)
             b_idx = _get_check_index(self.transb, True,
                                      row_idx=-2, col_idx=-1)
             type_check.expect(
+                a_type.ndim == b_type.ndim,
                 a_type.shape[:-2] == b_type.shape[:-2],
                 a_type.shape[a_idx] == b_type.shape[b_idx],
             )
 
     def forward(self, x):
         self.retain_inputs((0, 1))
         a, b = x
-        y = _matmul(a, b, self.transa, self.transb)
+        if a.ndim == 0 or b.ndim == 0:
+            y = a * b
+        else:
+            y = _matmul(a, b, self.transa, self.transb, self.transc)
         return utils.force_array(y),
 
     def backward(self, indexes, grad_outputs):
         a, b = self.get_retained_inputs()
-        return MatMulGrad(self.transa, self.transb).apply(
-            (a, b, grad_outputs[0]))
-
-
-class MatMulGrad(function_node.FunctionNode):
+        gy, = grad_outputs
 
-    def __init__(self, transa=False, transb=False):
-        self.transa = transa
-        self.transb = transb
+        ga = None
+        if 0 in indexes:
+            ga, = MatMul(self.transc, not self.transb,
+                         self.transa).apply((gy, b))
+            ga.data = ga.data.astype(a.dtype)
```

Comment:
Pelase use `chainer.functions.cast`.

Reply:
Well, I think we don't need to use a function for data type conversion here.

If we use `chainer.functions.cast`, inverse data type conversion is to be done by backward() of cast, when you double-backprop it. However, I think `MatMul` handles that data conversion in it, so I don't think we need to use `chainer.function.cast` here.

What would you think on this?

Reply:
Assuming there is a network that computes `c` = MatMul(`a`, `b`) in it and dtype of `a`, `b` and `c` are fp16, fp32 and fp32, respectively (I'm not sure whether this combination of data type is practical or not...). When you backprop it, `ga` will be computed by something like `ga` = MatMul(`gc`, `b`). A data type of `ga` becomes fp32, because data type of `gc` and `b` are fp32. However data type of `a` and `ga` must be the same. That's why data type conversion for `ga` (from fp32 to fp16) is necessary here.

Here is the main issue. If you use a function (`chainer.functions.cast`) for this data type conversion, what would happen when you double-backprop it?  `ggc` will be computed by something like `ggc` = MatMul(`gga`, `b`). I think the MatMul for double-backprop should handle the same combination of inputs in terms of data type as the MatMul for forward, say `c` = MatMul(`a`, `b`). However, data type of `gga` gets fp32 thanks to backward() of `chainer.functins.cast`. If we don't use a function here, `gga` becomes fp16.  I think is is better in consistency point of view.

What would you think?

Reply:
OK, how about introduce `dtype` option to `MatMul` function?
And, `astype` has a `copy` option. Please set `copy=False`.

Reply:
That's good idea! will fix in that way.